### PR TITLE
Better display of Nullables

### PR DESCRIPTION
**Pull request by amellnik**
As per #1082

### CODE REVIEW

**Comment by amellnik** [comment]
As per #1082

File: src/abstractdataframe/io.jl [hunk 1]
```diff
@@ -184,8 +184,13 @@ end
         write(io, "<tr>")
         write(io, "<th>$row</th>")
         for column_name in cnames
-            cell = string(df[row, column_name])
+          if isnull(df[row, column_name])
```
Comment (nalimilan) [inline src/abstractdataframe/io.jl]: You don't need this. Just call [`ourshowcompact`](https://github.com/amellnik/DataFrames.jl/blob/baeb0206fa3e0ac87fecd04d4fd48c5161e0d854/src/abstractdataframe/show.jl#L64) just like (`show`)[https://github.com/amellnik/DataFrames.jl/blob/baeb0206fa3e0ac87fecd04d4fd48c5161e0d854/src/abstractdataframe/show.jl#L255] does. Note in particular the `try...catch` block for undef entries.

Also, please use four-space indents just like in the rest of the code.

Comment (amellnik) [inline src/abstractdataframe/io.jl]: I tried `ourshowcompact` but I think I might be missing something as I'm still seeing the quotes:
image

**Comment by nalimilan** [comment]
Yeah, that's how it's printed at the REPL. I think it's good to be consistent, but we can discuss whether we should print the quotes at the REPL too. Looks like it's common in data base tools, including R and Pandas data frames, to omit quotes around strings. And indeed our docs for `outshowcompact` say that quotes are supposed to be omitted. I think you can just change the method for `AbstractString` to use `print` instead of `showcompact`, and it should be good. Be sure this is tested too.

I'll need to find how to make this work well for `CategoricalArray{String}`, but that's for another PR.

**Comment by amellnik** [comment]
Both `print` and `showcompact` include the quotation marks, so I added stripping them to a `Nullable{String}` method for `ourshowcompact`.  If this approach is OK let me know, and I'll add tests and squash.

**Comment by nalimilan** [comment]
Please avoid special-casing as much as possible. For this, make your code as similar as possible to `show` (see the link I provided above). The different printing methods shouldn't diverge unless there's a real reason to. So please get rid of the `isnull` branch, which isn't needed since `ourshowcompact` works for null values already.

File: src/abstractdataframe/show.jl [hunk 2]
```diff
@@ -64,6 +64,7 @@ end
 ourshowcompact(io::IO, x::Any) = showcompact(io, x) # -> Void
 ourshowcompact(io::IO, x::AbstractString) = showcompact(io, x) # -> Void
 ourshowcompact(io::IO, x::Symbol) = print(io, x) # -> Void
+ourshowcompact(io::IO, x::Nullable{String}) = isnull(x) ? print(io, "NA") : print(io, replace(get(x),  r"^\"|\"$", "")) # -> Void
```
Comment (nalimilan) [inline src/abstractdataframe/show.jl]: Much simpler and cleaner: `isnull(x) ? showcompact(io, x) : print(io, get(x))`.

File: src/abstractdataframe/show.jl [hunk 3]
```diff
@@ -64,6 +64,7 @@ end
 ourshowcompact(io::IO, x::Any) = showcompact(io, x) # -> Void
 ourshowcompact(io::IO, x::AbstractString) = showcompact(io, x) # -> Void
```
Comment (nalimilan) [inline src/abstractdataframe/show.jl]: Need to change `showcompact` to `print` if you want to remove quotes.

File: test/io.jl [hunk 4]
```diff
@@ -490,4 +490,11 @@ module TestIO
     # Need to wrap macro call inside eval to prevent the error from being
     # thrown prematurely
     @test_throws ArgumentError eval(:(csv"foo,bar"a))
+
+    #Test HTML output for IJulia and similar
+    df = DataFrame(Fish = ["Suzy", "Amir"], Mass = [1.5, Nullable{Float64}()])
+    io = IOBuffer()
+    show(io, "text/html", df)
+    str = takebuf_string(io)
+    @test str == "<table class=\"data-frame\"><tr><th></th><th>Fish</th><th>Mass</th></tr><tr><th>1</th><td>Suzy</td><td>1.5</td></tr><tr><th>2</th><td>Amir</td><td>#NULL</td></tr></table>"
```
Comment (nalimilan) [inline test/io.jl]: Can you break this into one line per row via `"""`?

File: test/show.jl [hunk 5]
```diff
@@ -35,4 +36,11 @@ module TestShow
     show(io, A)
     A = DataFrames.RepeatedVector([1, 2, 3], 1, 5)
     show(io, A)
+
+    #Test show output for REPL and similar
+    df = DataFrame(Fish = ["Suzy", "Amir"], Mass = [1.5, Nullable{Float64}()])
+    io = IOBuffer()
+    show(io, df)
+    str = takebuf_string(io)
+    @test str == "2×2 DataFrames.DataFrame\n│ Row │ Fish │ Mass  │\n├─────┼──────┼───────┤\n│ 1   │ Suzy │ 1.5   │\n│ 2   │ Amir │ #NULL │"
```
Comment (nalimilan) [inline test/show.jl]: Same here.

File: test/io.jl [hunk 6]
```diff
@@ -490,4 +490,11 @@ module TestIO
     # Need to wrap macro call inside eval to prevent the error from being
     # thrown prematurely
     @test_throws ArgumentError eval(:(csv"foo,bar"a))
+
+    #Test HTML output for IJulia and similar
+    df = DataFrame(Fish = ["Suzy", "Amir"], Mass = [1.5, Nullable{Float64}()])
```
Comment (nalimilan) [inline test/io.jl]: FWIW, you don't actually need to specify `{Float64}` since promotion does that for you.

File: test/io.jl (hunk 4 above)
Comment (amellnik) [inline test/io.jl]: On this one, the HTML output doesn't contain any linebreaks.  Is there any easy way to compare the strings ignoring `\n`?  The REPL version doesn't have this issue.

Comment (nalimilan) [inline test/io.jl]: Ah, indeed. Don't worry about it then. Just rebase and it should be good to go.

**Comment by nalimilan** [comment]
Thanks, but you still need to rebase against latest master.

**Comment by nalimilan** [comment]
Test failed because of an extraneous ending `\n`.

**Comment by amellnik** [comment]
Fixed my test.  The Latex test was failing because the Latex output was writing lines with trailing whitespace.  When I edited io.jl in Atom this whitespace was automatically trimmed, leading the test to fail.  I don't think the trailing whitespace was intentional, so I removed it and tests now pass.

File: src/abstractdataframe/io.jl [hunk 7]
```diff
@@ -232,8 +232,8 @@ function Base.show(io::IO, ::MIME"text/latex", df::AbstractDataFrame)
     write(io, "\t& ")
     header = join(map(c -> latex_escape(string(c)), cnames), " & ")
     write(io, header)
-    write(io, "\\\\ \n")
```
Comment (nalimilan) [inline src/abstractdataframe/io.jl]: @maximerischard Was the space intended?

**Comment by nalimilan** [comment]
Thanks, sorry for the delay.

I suspect we'll have to add another special case for CategoricalArray printing. If you want to give it a look...

File: src/abstractdataframe/io.jl (hunk 7 above)
Comment (maximerischard) [inline src/abstractdataframe/io.jl]: I think it was just to make it a bit more readable than having 5 backslashes in a row.

**Pull request merged**